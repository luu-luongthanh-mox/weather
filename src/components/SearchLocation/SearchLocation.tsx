import React, { useEffect, useState } from "react";
import { Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { AsyncTypeahead } from "react-bootstrap-typeahead";
import { toast } from "react-toastify";

import { search } from "@stores/location";
import { getLocations } from "@stores/location/selector";
import { useDebounce } from "@hooks/useDebounce";
import { SearchResult } from "@models/SearchResult";
import { StateStatus } from "@models/State";

interface SearchLocationProps {
  onLocationChanged: (selected: SearchResult) => void;
}
export const SearchLocation: React.FC<SearchLocationProps> = ({
  onLocationChanged,
}) => {
  const { error, loading, searchResult } = useSelector(getLocations);
  const dispatch = useDispatch();

  if (error) {
    toast.error(error);
  }

  const [searchText, setSearchText] = useState("");
  const debouncedSearchText = useDebounce(searchText, 500);

  useEffect(() => {
    dispatch(search(debouncedSearchText));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [debouncedSearchText]);

  const onSearchChange = (text: string) => {
    setSearchText(text);
  };

  const onSelected = (selected: SearchResult[]) => {
    onLocationChanged(selected[0]);
  };

  return (
    <Form>
      <Form.Group controlId="formBasicEmail">
        <Form.Label>Search City</Form.Label>
        <AsyncTypeahead
          isLoading={loading === StateStatus.loading}
          onChange={onSelected}
          onSearch={onSearchChange}
          options={searchResult}
          clearButton={true}
          id="searchLocation"
        />
      </Form.Group>
    </Form>
  );
};
