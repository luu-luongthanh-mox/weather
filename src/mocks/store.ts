import { RootState } from "@stores";
import { StateStatus } from "@models/State";

import { forecastWeatherItemMock } from "./forecastWeather";

export const storeMock: RootState = {
  forecastWeather: {
    data: {
      consolidated_weather: [forecastWeatherItemMock],
      time: "2020-12-06T10:54:45.650562+08:00",
      sun_rise: "2020-12-06T06:50:25.886047+08:00",
      sun_set: "2020-12-06T17:39:03.353222+08:00",
      timezone_name: "LMT",
      title: "Hong Kong",
      location_type: "City",
      woeid: 2165352,
      latt_long: "22.411200,114.153999",
      timezone: "Asia/Hong_Kong",
    },
    loading: StateStatus.idle,
    currentRequestId: undefined,
  },
  location: {
    loading: StateStatus.idle,
    selected: null,
    searchResult: [],
    currentRequestId: undefined,
  },
};
