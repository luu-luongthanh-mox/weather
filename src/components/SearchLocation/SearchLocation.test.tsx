import React from "react";
import { Provider } from "react-redux";
import createMockStore from "redux-mock-store";
import { render } from "@testing-library/react";
import { getDefaultMiddleware } from "@reduxjs/toolkit";

import { SearchLocation } from "./SearchLocation";
import { storeMock } from "@mocks/store";

const mockStore = createMockStore(getDefaultMiddleware());
const mockOnLocationChanged = jest.fn();

describe("Component: SearchLocation", () => {
  it("renders data on screen", () => {
    const store = mockStore(storeMock);
    const { asFragment } = render(
      <Provider store={store}>
        <SearchLocation onLocationChanged={mockOnLocationChanged} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
