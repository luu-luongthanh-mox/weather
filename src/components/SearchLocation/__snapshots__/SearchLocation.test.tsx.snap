// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SearchLocation renders data on screen 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="formBasicEmail"
      >
        Search City
      </label>
      <div
        class="rbt"
        style="outline: none; position: relative;"
        tabindex="-1"
      >
        <div
          style="display: flex; flex: 1; height: 100%; position: relative;"
        >
          <input
            aria-autocomplete="both"
            aria-expanded="false"
            aria-haspopup="listbox"
            autocomplete="off"
            class="rbt-input-main form-control rbt-input"
            role="combobox"
            type="text"
            value=""
          />
          <input
            aria-hidden="true"
            class="rbt-input-hint"
            readonly=""
            style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.35); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%;"
            tabindex="-1"
            value=""
          />
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
